### ISSUE TITLE
Bayesian MCMC

### ISSUE THREAD

**Issue by jbogaardt** [jbogaardt/chainladder-python#30]
Add method to do Baesian MCMC.  See:
[CAS Monograph #1](https://www.casact.org/pubs/monographs/index.cfm?fa=meyers-monograph01)

**Comment by danielfong-act** [comment]
Saw this on the agenda today, and coincidentally I have been looking at implementing this. It seems like a good time to post what I've got.

First, I found a bug in the monograph's LCL/CCL code. If it's not just an issue on my machine, then you should get a bunch of NAs in the results when you run the code. CCL code sheet line/cell 174/175 should read
```
rho=b[,30]
sigd2=b[,20:29]
```

I've been using pymc/pytensor, which let you drop triangles in as pandas frames or numpy arrays. For the data, clrd doesn't match the csv on the CAS site (used by the paper), so I manually pulled it from the csv in the same manner as the R code. For the NUTS sampler, I use blackjax, which can be invoked in calls to `sample()`. This was the easiest way I found to get things done and keep everything in python. Using pymc otherwise, without g++, is intolerably slow. Also the progress bar initially worked for me but now doesn't, so I turned it off (in `sample()`).
<details>
<summary>Here's what I've got to replicate the LCL model from the paper/spreadsheet</summary>

```import pymc as pm, pytensor as pt
import chainladder as cl, pandas as pd, numpy as np

celinadata = pd.read_csv(r'C:\users\danie\downloads\comauto_pos.csv')
celinadata = celinadata[celinadata['GRCODE']==353]
celinadata = celinadata[(celinadata['AccidentYear']+celinadata['DevelopmentLag'])<1999] 
celinadata['incloss']=celinadata['IncurLoss_C']-celinadata['BulkLoss_C']
celinatri = cl.Triangle(celinadata,origin='AccidentYear',development='DevelopmentYear',columns=['incloss','EarnedPremNet_C'],cumulative=True)
tri  = celinatri['incloss']
sample_weight =  celinatri['EarnedPremNet_C'].latest_diagonal
  
model = pm.Model()
wlim = len(tri.origin)
dlim = len(tri.development)
with model:
    
    aprior = pm.Uniform('a',.000001,1,shape=(dlim,))
    sigmaprior = pm.Deterministic('sigma',pm.math.stack([sum(aprior[i:dlim]) for i in range(dlim)])) # decreasing sigma for LCL/CCL
    logelrprior = pm.Uniform('logelr',-1.5,.5) 
    alphaprior = pm.Normal('alpha',mu=np.log(sample_weight).to_frame().values.T[0]+logelrprior,sigma=np.sqrt(10))
    betacoef = ([-5]*dlim,[5]*dlim)
    betacoef[0][dlim-1] = -.000001
    betacoef[1][dlim-1] = .000001
    betaprior = pm.Uniform('beta',betacoef[0],betacoef[1]) # last beta = 0 
    # rhoprior = pm.Uniform.dist(-1,1) 
    muprior = pm.Deterministic('mu',alphaprior[:,None] + betaprior[None,:])
    C =  pm.Normal('Cumulative',mu=muprior,sigma=sigmaprior,observed=tri.log().to_frame())
    
    
    Cpred=pm.sample(5000,nuts_sampler='blackjax',progressbar=False)
    print(np.exp(Cpred.posterior.Cumulative[:,:,:,9]).mean(axis=0).mean(axis=0))

```

</details>

Things seem straightforward, but when I go to sample I get several hundred divergences and large rhat stats, indicating that something went wrong. I'm working on it, but my intro bayesian stats course didn't cover HMC so some of it's a bit over my head. I also might've misspecified the model somewhere

**Comment by genedan** [comment]
@danielfong-act, I'm assuming you're still interested in this, so I've labeled you as an assignee.

I came across this repo today, maybe it could be of use?

https://github.com/DrPeterEngland/StochasticReserving/tree/main

Cc'ing @DrPeterEngland - I'm not sure if you have encountered us before, so I'll give a short intro. This is [chainladder-python](https://chainladder-python.readthedocs.io/), a package that strives to incorporate reserving methods from the actuarial community. We're open to the general public, and I hope we've catched your attention, as your package has ours - thanks for putting your repo out there!

**Comment by DrPeterEngland** [comment]
Hi,

Many thanks for your message. Yes, I did find chainladder-python last October while I was developing my functions. I was looking for something broadly equivalent to the ChainLadder package in R. Specifically I was looking for something that:


  *   calculated Mack’s standard errors analytically
  *   calculated the standard deviation of the claims development result for a sequence of one-year views to ultimate

chainladder-python is able to do the former, but not the latter (as far as I can see).

I was just starting out with Python, so I struggled a bit to get going quickly with chainladder-python. Given a claims triangle imported as a simple csv file, I had to convert the triangle to a pandas dataframe in long format then convert to dates before converting back to a triangle that chainladder-python would accept. I can see why using dates etc and a long format is a good thing for a production system, but for me as a novice it was a bit of a barrier to entry. However, I was able to find the information needed to get me going, which was great.

When applying Mack’s model, I wanted to set the final sigma value to the minimum of the previous two values. I found a way to do that, which again was pleasing.

My initial attempts at the Jupyter examples did in fact include an example using chainladder-python for Mack’s standard errors calculated analytically, as well as an example using my own function fitting a weighted normal GLM and calculating the standard errors using the recursive formulae in the appendix of England & Verrall (2002). In the end I dropped the chainladder-python example (apologies!) since it could not calculate the standard deviation of the claims development result for a sequence of one-year views to ultimate. I didn’t create a function for that either, so it’s not available in Python yet as far as I know.

I’ll have another look at chainladder-python when I have a bit more time, and thanks for your contribution.

Kind regards,

Peter

Peter England, PhD
Director
EMC Actuarial & Analytics

Mobile: +44 7941 457974
Email: ***@***.******@***.***>

[Logo 3]

**Comment by danielfong-act** [comment]
Hey @genedan, sorry for the late update. I really hit a wall trying to get the most basic of the models described in the monograph (LCL) to work. I'm gonna take another crack at it, and if I don't make any real progress soon, then I will put up some framework at least to show the direction I was trying to move in. I think pymc is a reasonable choice for implementing bayesian models in general, but pystan (which the project you linked uses) would probably also work fine. 

I haven't read the 06 paper from that project (paywall), but [this paper](https://forum.casact.org/article/122945-handling-sparse-data-for-reserving-using-bayesian-mcmc) specifies a model completely different from those of the monograph. Thus, I think that eventually chainladder should support a very wide family of bayesian models. I know pymc would easily allow for this flexibility, but I can't say one way or the other about other packages. 

Anyways that's where I'm at. I'll keep you posted

**Comment by genedan** [comment]
Yeah no worries, we can definitely use all the help we can get. Thanks!
